app/app.go

```diff
@@ -163,14 +172,35 @@ func (app *Basecoin) BeginBlock(hash []byte, header *abci.Header) {
 }
 
 // EndBlock - ABCI
+// Returns a list of all validator changes made in this block
 func (app *Basecoin) EndBlock(height uint64) (res abci.ResponseEndBlock) {
-	// for _, plugin := range app.plugins.GetList() {
-	// 	pluginRes := plugin.EndBlock(app.state, height)
-	// 	res.Diffs = append(res.Diffs, pluginRes.Diffs...)
-	// }
+	// TODO: cleanup in case a validator exists multiple times in the list
+	res.Diffs = app.pending
+	app.pending = nil
 	return
 }
 
+func (app *Basecoin) addValChange(diffs []*abci.Validator) {
+	for _, d := range diffs {
+		idx := findVal(d, app.pending)
+		if idx >= 0 {
+			app.pending[idx] = d
+		} else {
+			app.pending = append(app.pending, d)
+		}
+	}
+}
+
+// return index of list with validator of same PubKey, or -1 if no match
+func findVal(val *abci.Validator, list []*abci.Validator) int {
```

Comment:
Perhaps we can change this name to: `pubKeyIndex`
so
```go
// pubKeyIndex returns the index of the item with the same PubKey
// as query, otherwise it returns -1 if no match is found.
func pubKeyIndex(query *abci.Validator, items []*abci.Validator) int
```